util/entropy.c

```diff
@@ -16,61 +18,157 @@
  * XXX win32 where there is no concept of /dev/urandom.
  */
 
+struct entropy_read
+{
+	int fd;
+};
+
 /**
- * entropy_read(buf, buflen):
- * Fill the given buffer with random bytes provided by the operating system.
+ * entropy_read_open():
+ * Initialize the ability to produce random bytes, and return a cookie.
  */
-int
-entropy_read(uint8_t * buf, size_t buflen)
+static struct entropy_read *
+entropy_read_open(void)
 {
-	int fd;
-	ssize_t lenread;
+	static struct entropy_read * er;
 
-	/* Sanity-check the buffer size. */
-	if (buflen > SSIZE_MAX) {
-		warn0("Programmer error: "
-		    "Trying to read insane amount of random data: %zu",
-		    buflen);
+	/* Allocate cookie. */
+	if ((er = malloc(sizeof(struct entropy_read))) == NULL) {
+		warnp("malloc");
 		goto err0;
 	}
 
 	/* Open /dev/urandom. */
-	if ((fd = open("/dev/urandom", O_RDONLY)) == -1) {
+	if ((er->fd = open("/dev/urandom", O_RDONLY)) == -1) {
 		warnp("open(/dev/urandom)");
+		goto err1;
+	}
+
+	/* Success! */
+	return (er);
+
+err1:
+	free(er);
+err0:
+	/* Failure! */
+	return (NULL);
+}
+
+/**
+ * entropy_read_fill(er, buf, buflen):
+ * Fill the given buffer with random bytes provided by the operating system
+ * using the file descriptor in ${er}.
+ */
+static int
+entropy_read_fill(struct entropy_read * er, uint8_t * buf, size_t buflen)
+{
+	ssize_t lenread;
+
+	/* Sanity checks. */
+	assert(er != NULL);
+	if (buflen > SSIZE_MAX) {
+		warn0("Programmer error: "
+		    "Trying to read insane amount of random data: %zu",
+		    buflen);
 		goto err0;
 	}
 
 	/* Read bytes until we have filled the buffer. */
 	while (buflen > 0) {
-		if ((lenread = read(fd, buf, buflen)) == -1) {
+		if ((lenread = read(er->fd, buf, buflen)) == -1) {
 			warnp("read(/dev/urandom)");
-			goto err1;
+			goto err0;
 		}
 
 		/* The random device should never EOF. */
 		if (lenread == 0) {
 			warn0("EOF on /dev/urandom?");
-			goto err1;
+			goto err0;
 		}
 
 		/* We've filled a portion of the buffer. */
 		buf += (size_t)lenread;
 		buflen -= (size_t)lenread;
 	}
 
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/**
+ * entropy_read_close(er):
+ * Close any file descriptor in ${er} and free memory.
```

Comment:
I would say "resources" rather than calling out "file descriptor" and "memory" specifically; a future Windows version of this might obtain a Handle, for example.

Reply:
Hmm, by "memory", I was referring to `struct entropy_read * er` itself.  I've now got
`Release any resources used by {er}.` which sounds a bit too vague to me, but OTOH it's probably good enough for the "clean-up" function?